Can't build erlang 29-rc1 with --enable-deterministic-build

**Describe the bug**

Looks like when configuring and building Erlang/OTP 29rc1 with `--enable-deterministic-build`, the makefile tooling can not create deps/deps.mk files having relative paths for .hrl dependencies.

**To Reproduce**

./configure --enable-deterministic-build



And then `make` it will give:

```
make[3]: Leaving directory '/home/joaohf/work/opensource/otp-29-rc1/erts/lib_src'
(cd preloaded/src && make ../../ebin/erts.app)
make[3]: Entering directory '/home/joaohf/work/opensource/otp-29-rc1/erts/preloaded/src'
erlc -M +debug_info +warn_missing_doc_function +warn_missing_doc_callback +warn_missing_spec_documented +deterministic -I/home/joaohf/work/opensource/otp-29-rc1/lib/kernel/src -I/home/joaohf/work/opensource/otp-29-rc1/lib/kernel/include +nowarn_deprecated_catch +deterministic -o ../../ebin erl_prim_loader.erl init.erl prim_buffer.erl prim_file.erl prim_inet.erl socket_registry.erl prim_socket.erl prim_net.erl zlib.erl prim_zip.erl erl_init.erl erts_code_purger.erl erlang.erl erts_internal.erl erl_tracer.erl erts_literal_area_collector.erl erts_trace_cleaner.erl erts_dirty_process_signal_handler.erl atomics.erl counters.erl persistent_term.erl \
| perl -pe "s@ [a-zA-Z]?/home/joaohf/work/opensource/otp-29-rc1/(?:bootstrap/)?lib/([^/]+)@ ../../../lib/\1@g" 2> /dev/null \
> deps/deps.mk
/bin/sh: 1: erlc: not found
mkdir "../../ebin"
sed -e 's;%VSN%;17.0;' -e 's;%ESOCK_MODS%;prim_net, prim_socket, socket_registry,;' erts.app.src > ../../ebin/erts.app
make[3]: Leaving directory '/home/joaohf/work/opensource/otp-29-rc1/erts/preloaded/src'
make[2]: Leaving directory '/home/joaohf/work/opensource/otp-29-rc1/erts'
make[1]: Leaving directory '/home/joaohf/work/opensource/otp-29-rc1/erts'
mkdir /home/joaohf/work/opensource/otp-29-rc1/bootstrap/bin/x86_64-pc-linux-gnu
cd lib && \
  ERL_TOP=/home/joaohf/work/opensource/otp-29-rc1 PATH=/home/joaohf/work/opensource/otp-29-rc1/bootstrap/bin:"${PATH}" \
        make opt BOOTSTRAP="parsetools sasl asn1/src" ERL_COMPILE_WARNINGS_AS_ERRORS=yes
make[1]: Entering directory '/home/joaohf/work/opensource/otp-29-rc1/lib'
make[2]: Entering directory '/home/joaohf/work/opensource/otp-29-rc1/lib/parsetools'
=== Entering application parsetools
make[3]: Entering directory '/home/joaohf/work/opensource/otp-29-rc1/lib/parsetools/src'
erlc -M -Werror +debug_info +warn_missing_doc_function +warn_missing_doc_callback +warn_missing_spec_documented +deterministic -I/home/joaohf/work/opensource/otp-29-rc1/lib/stdlib/include -Werror +warn_obsolete_bool_op +nowarn_deprecated_catch -o ../ebin leex.erl yecc.erl yeccscan.erl \
| perl -pe "s@ [a-zA-Z]?/home/joaohf/work/opensource/otp-29-rc1/(?:bootstrap/)?lib/([^/]+)@ ../../\1@g" 2> /dev/null \
> deps/deps.mk
make[3]: *** No rule to make target 'erl_compile.hrl', needed by '../ebin/leex.beam'.  Stop.
make[3]: Leaving directory '/home/joaohf/work/opensource/otp-29-rc1/lib/parsetools/src'
make[2]: *** [/home/joaohf/work/opensource/otp-29-rc1/make/otp_subdir.mk:31: opt] Error 2
make[2]: Leaving directory '/home/joaohf/work/opensource/otp-29-rc1/lib/parsetools'
make[1]: *** [/home/joaohf/work/opensource/otp-29-rc1/make/otp_subdir.mk:31: opt] Error 2
make[1]: Leaving directory '/home/joaohf/work/opensource/otp-29-rc1/lib'
make: *** [Makefile:683: secondary_bootstrap] Error 2
```

I think this issue is related to this commit:

```
b2563f1eb5 ("otp: Fix makefile deps with erlc -M for almost all application", 2025-05-15)
```

**Expected behavior**
Build works with --enable-deterministic-build flag as it used to work as before.

**Affected versions**
OTP 29-rc1

**Additional context**

The following is from a yocto build (but I tested the above scenario outside yocto context too):

```
| make[3]: Entering directory '/build/tmp/work/x86_64-linux/erlang-native/29.0-rc1/sources/erlang-29.0-rc1/erts/preloaded/src'
| erlc -M +debug_info +warn_missing_doc_function +warn_missing_doc_callback +warn_missing_spec_documented +deterministic -I/build/tmp/work/x86_64-linux/erlang-native/29.0-rc1/sources/erlang-29.0-rc1/lib/kernel/src -I/build/tmp/work/x86_64-linux/erlang-native/29.0-rc1/sources/erlang-29.0-rc1/lib/kernel/include +nowarn_deprecated_catch +deterministic -o ../../ebin erl_prim_loader.erl init.erl prim_buffer.erl prim_file.erl prim_inet.erl socket_registry.erl prim_socket.erl prim_net.erl zlib.erl prim_zip.erl erl_init.erl erts_code_purger.erl erlang.erl erts_internal.erl erl_tracer.erl erts_literal_area_collector.erl erts_trace_cleaner.erl erts_dirty_process_signal_handler.erl atomics.erl counters.erl persistent_term.erl \
| | perl -pe "s@ [a-zA-Z]?/build/tmp/work/x86_64-linux/erlang-native/29.0-rc1/sources/erlang-29.0-rc1/(?:bootstrap/)?lib/([^/]+)@ ../../../lib/\1@g" 2> /dev/null \
| > deps/deps.mk
| /bin/sh: 1: erlc: not found
| mkdir "../../ebin"
| sed -e 's;%VSN%;17.0;' -e 's;%ESOCK_MODS%;prim_net, prim_socket, socket_registry,;' erts.app.src > ../../ebin/erts.app
| make[3]: Leaving directory '/build/tmp/work/x86_64-linux/erlang-native/29.0-rc1/sources/erlang-29.0-rc1/erts/preloaded/src'
| make[2]: Leaving directory '/build/tmp/work/x86_64-linux/erlang-native/29.0-rc1/sources/erlang-29.0-rc1/erts'
| make[1]: Leaving directory '/build/tmp/work/x86_64-linux/erlang-native/29.0-rc1/sources/erlang-29.0-rc1/erts'
| mkdir /build/tmp/work/x86_64-linux/erlang-native/29.0-rc1/sources/erlang-29.0-rc1/bootstrap/bin/x86_64-pc-linux-gnu
| cd lib && \
|   ERL_TOP=/build/tmp/work/x86_64-linux/erlang-native/29.0-rc1/sources/erlang-29.0-rc1 PATH=/build/tmp/work/x86_64-linux/erlang-native/29.0-rc1/sources/erlang-29.0-rc1/bootstrap/bin:"${PATH}" \
|       make opt BOOTSTRAP="parsetools sasl asn1/src" ERL_COMPILE_WARNINGS_AS_ERRORS=yes
| make[1]: Entering directory '/build/tmp/work/x86_64-linux/erlang-native/29.0-rc1/sources/erlang-29.0-rc1/lib'
| make[2]: Entering directory '/build/tmp/work/x86_64-linux/erlang-native/29.0-rc1/sources/erlang-29.0-rc1/lib/parsetools'
| === Entering application parsetools
| make[3]: Entering directory '/build/tmp/work/x86_64-linux/erlang-native/29.0-rc1/sources/erlang-29.0-rc1/lib/parsetools/src'
| erlc -M -Werror +debug_info +warn_missing_doc_function +warn_missing_doc_callback +warn_missing_spec_documented +deterministic -I/build/tmp/work/x86_64-linux/erlang-native/29.0-rc1/sources/erlang-29.0-rc1/lib/stdlib/include -Werror +warn_obsolete_bool_op +nowarn_deprecated_catch -o ../ebin leex.erl yecc.erl yeccscan.erl \
| | perl -pe "s@ [a-zA-Z]?/build/tmp/work/x86_64-linux/erlang-native/29.0-rc1/sources/erlang-29.0-rc1/(?:bootstrap/)?lib/([^/]+)@ ../../\1@g" 2> /dev/null \
| > deps/deps.mk
| make[3]: Leaving directory '/build/tmp/work/x86_64-linux/erlang-native/29.0-rc1/sources/erlang-29.0-rc1/lib/parsetools/src'
| make[3]: *** No rule to make target 'erl_compile.hrl', needed by '../ebin/leex.beam'.  Stop.
| make[2]: *** [/build/tmp/work/x86_64-linux/erlang-native/29.0-rc1/sources/erlang-29.0-rc1/make/otp_subdir.mk:31: opt] Error 2
| make[2]: Leaving directory '/build/tmp/work/x86_64-linux/erlang-native/29.0-rc1/sources/erlang-29.0-rc1/lib/parsetools'
| make[1]: *** [/build/tmp/work/x86_64-linux/erlang-native/29.0-rc1/sources/erlang-29.0-rc1/make/otp_subdir.mk:31: opt] Error 2
| make[1]: Leaving directory '/build/tmp/work/x86_64-linux/erlang-native/29.0-rc1/sources/erlang-29.0-rc1/lib'
```

And the following patch tries to solve the situation:

```
diff --git a/make/dep.mk b/make/dep.mk
index 2da7bf24d4..766b95279a 100644
--- a/make/dep.mk
+++ b/make/dep.mk
@@ -32,11 +32,13 @@ $(shell mkdir -p $(dir $(DEP_FILE)) >/dev/null)
 
 DEP_REL_TOP?=../../
 
+DEP_FILTER_ERL_COMPILE_FLAGS=+deterministic
+
 deps: $(DEP_FILE)
 
 $(DEP_FILE): $(ERL_FILES) $(HRL_FILES) $(INTERNAL_HRL_FILES) $(EXTRA_DEP_DEPENDENCIES) $(ERL_TOP)/make/dep.mk Makefile
-       $(gen_verbose)erlc -M $(ERL_COMPILE_FLAGS) -o $(EBIN) $(filter-out $(DEP_SKIP_FILES), $(ERL_FILES)) \
+       $(gen_verbose)erlc -M $(filter-out $(DEP_FILTER_ERL_COMPILE_FLAGS), $(ERL_COMPILE_FLAGS)) -o $(EBIN) $(filter-out $(DEP_SKIP_FILES), $(ERL_FILES)) \
        | perl -pe "s@ [a-zA-Z]?$(ERL_TOP_NATIVE)/(?:bootstrap/)?lib/([^/]+)@ $(DEP_REL_TOP)\1@g" 2> /dev/null \
        > $(DEP_FILE)
 
--include $(DEP_FILE)
\ No newline at end of file
+-include $(DEP_FILE)
```
But it will not work for cases where there is a call like this:

```
erlc -W  +deterministic -o. erl_parse.yrl
```

Where erl_parse.beam will be deterministic.

<!--

%CopyrightBegin%

SPDX-License-Identifier: Apache-2.0

Copyright Ericsson AB 2021-2025. All Rights Reserved.

%CopyrightEnd%

-->

--- a/lib/compiler/src/Makefile
+++ b/lib/compiler/src/Makefile
@@ -145,6 +145,7 @@ ERL_COMPILE_FLAGS += +inline +warn_unused_import \
 
 ifeq ($(ERL_DETERMINISTIC),yes)
 	DETERMINISM_FLAG = +deterministic
+	DEP_SKIP_FILES = core_parse.erl
 else
 	DETERMINISM_FLAG =
 endif

--- a/lib/edoc/src/Makefile
+++ b/lib/edoc/src/Makefile
@@ -51,6 +51,10 @@ XMERL = ../../xmerl
 ERL_COMPILE_FLAGS += -pa $(XMERL) -I../include -I$(XMERL)/include -I${ERL_TOP}/lib
 ERL_COMPILE_FLAGS += +warn_unused_vars +warn_unused_import +warn_deprecated_guard +no_docs +nowarn_missing_spec_documented +warn_deprecated_catch -Werror
 
+ifeq ($(ERL_DETERMINISTIC),yes)
+	DEP_SKIP_FILES = edoc_parser.erl
+endif
+
 include files.mk
 
 OBJECTS=$(ERL_FILES:%.erl=$(EBIN)/%.$(EMULATOR)) $(APP_TARGET) $(APPUP_TARGET)

--- a/lib/stdlib/src/Makefile
+++ b/lib/stdlib/src/Makefile
@@ -183,6 +183,7 @@ ERL_COMPILE_FLAGS += -I../include -I../../kernel/include
 
 ifeq ($(ERL_DETERMINISTIC),yes)
 	DETERMINISM_FLAG = +deterministic
+	DEP_SKIP_FILES = erl_parse.erl
 else
 	DETERMINISM_FLAG =
 endif

--- a/lib/tools/src/Makefile
+++ b/lib/tools/src/Makefile
@@ -82,6 +82,10 @@ CSS = $(PRIVDIR)/styles.css
 # ----------------------------------------------------
 ERL_COMPILE_FLAGS += -Werror +nowarn_deprecated_catch
 
+ifeq ($(ERL_DETERMINISTIC),yes)
+	DEP_SKIP_FILES = xref_parser.erl
+endif
+
 # ----------------------------------------------------
 # Targets
 # ----------------------------------------------------

--- a/lib/xmerl/src/Makefile
+++ b/lib/xmerl/src/Makefile
@@ -133,6 +133,14 @@ ERL_COMPILE_FLAGS += \
 
 ifeq ($(ERL_DETERMINISTIC),yes)
 	DETERMINISM_FLAG = +deterministic
+	DEP_SKIP_FILES = xmerl_b64Bin.erl \
+		xmerl_xpath_parse.erl \
+		xmerl_xsd_re_parse.erl \
+		xmerl_sax_parser_list.erl \
+		xmerl_sax_parser_latin1.erl \
+		xmerl_sax_parser_utf8.erl \
+		xmerl_sax_parser_utf16be.erl \
+		xmerl_sax_parser_utf16le.erl
 else
 	DETERMINISM_FLAG =
 endif

--- a/make/dep.mk
+++ b/make/dep.mk
@@ -30,12 +30,14 @@ DEPDIR=deps
 DEP_FILE=$(DEPDIR)/deps.mk
 $(shell mkdir -p $(dir $(DEP_FILE)) >/dev/null)
 
+DEP_FILTER_ERL_COMPILE_FLAGS=+deterministic
+
 DEP_REL_TOP?=../../
 
 deps: $(DEP_FILE)
 
 $(DEP_FILE): $(ERL_FILES) $(HRL_FILES) $(INTERNAL_HRL_FILES) $(EXTRA_DEP_DEPENDENCIES) $(ERL_TOP)/make/dep.mk Makefile
-	$(gen_verbose)erlc -M $(ERL_COMPILE_FLAGS) -o $(EBIN) $(filter-out $(DEP_SKIP_FILES), $(ERL_FILES)) \
+	$(gen_verbose)erlc -M $(filter-out $(DEP_FILTER_ERL_COMPILE_FLAGS), $(ERL_COMPILE_FLAGS)) -o $(EBIN) $(filter-out $(DEP_SKIP_FILES), $(ERL_FILES)) \
 	| perl -pe "s@ [a-zA-Z]?$(ERL_TOP_NATIVE)/(?:bootstrap/)?lib/([^/]+)@ $(DEP_REL_TOP)\1@g" 2> /dev/null \
 	> $(DEP_FILE)
 

